158 store user preferences as json - PR #1: Core JSON Preferences Foundation

eremozdemir opened the pull request:
<!--
Thank you for contributing! Please fill out this template to help us review your PR.
-->

## Description

Implemented the intial foundation for the `UserPreferences` class with `JSON` storage capabilities and integrates persistence into existing CLI commands. 

Methods added
1. `UserPreferences` core methods:
   - `load_preferences`
   - `save_preferences`
   - `update`
   - `get`
   - plus other getters/setters

2. CLI integration methods:
    - `_load_existing_preferences`
    - updated `do_perms`
    - `do_filepath`
    - `do_email`
    - `do_begin`

This automatically save and restore user data between sessions, ex `preferences.json`:
```
{
    "consent": false,
    "files_to_ignore": [],
    "file_start_time": null,
    "file_end_time": null,
    "last_updated": "2025-11-10T04:39:38.045395",
    "project_filepath": "",
    "user_name": "",
    "user_password": "",
    "user_email": ""
}
```


**Closes:** #158 **BUT NOT UNTIL ALL THREE PR'S ARE MERGED**

---

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [X] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Documentation added/updated
- [ ] Test added/updated **-> TBD in future PR**
- [ ] Refactoring
- [X] Performance improvement

---

## Testing

> Please describe how you tested this PR (both manually and with tests).
> Provide instructions so we can reproduce.

- [ ] Test A **-> TBD in future PR**
- [ ] Test B

---

## Checklist

- [ ] GenAI was used in generating the code and I have performed a self-review of my own code
- [X] I have commented my code where needed
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [X] Any dependent changes have been merged and published in downstream modules
- [X] Any UI changes have been checked to work on desktop, tablet, and/or mobile

---

## Screenshots

> If applicable, add screenshots to help explain your changes

<details>
<summary>Click to expand screenshots</summary>

<!-- Add your screenshots here -->

</details>

thndlovu reviewed (approved):
Solid foundation for user persistence, clean structure, useful core methods, and clear CLI integration. My only comment is adding tests cases, but if they’re planned for a future PR then LGTM.

sjsikora commented on src/classes/cli.py:
```diff
@@ -5,8 +5,110 @@
+    """Manages User Preferences with JSON storage in database folder"""
+
+    def __init__(self, preferences_file: str = None):
+        src_folder_path = Path(__file__).parent.parent
```
Is there a better way to find the database file? I am just thinking if we ever move this file around, we will run into an error.

sjsikora reviewed (changes requested):
Looks good! I just had a few comments, and if you could address them that would be great. Also, I didn't see any functionality for saving the preferences after they were populated by the user. I assume that that would come later as this is just the core foundation, right?

eremozdemir commented:
> Looks good! I just had a few comments, and if you could address them that would be great. Also, I didn't see any functionality for saving the preferences after they were populated by the user. I assume that that would come later as this is just the core foundation, right?

I resolved your comments (please see added changes in newest commit). For this first comment here is what I changed:

1) Here is what I’ve added for the Missing preference saving functionality:
a. In `do_perms`, from: 
```
            if answer == 'Y':  # user consents
                self.user_consent = True
                print("\nThank you for consenting. You may now continue.")
```
To:  
```
            if answer == 'Y':  # user consents
                self.user_consent = True
                # Save consent to preferences
                success = self.preferences.update_consent(True)
                if success:
                    print("\nThank you for consenting. Consent saved to preferences.")
                else:
                    print("\nConsent recorded but failed to save to preferences file.")
```
b. In `do_filepath`, from:
```
        # Process the filepath
        self.project_filepath = answer
        print("\nFilepath successfully received")
```
To:
```
        # Process the filepath
        self.project_filepath = answer
        # Save filepath to preferences
        success = self.preferences.update_project_filepath(answer)
        print("\nFilepath successfully received and saved to preferences")
```
c. In `do_email`, from:
```
        # Process the filepath
        self.user_email = answer
        print("\nEmail successfully received")
```
To:
```
        # Process the email
        self.user_email = answer
        # Save email to preferences
        success = self.preferences.update_user_email(answer)
        print("\nEmail successfully received and saved to preferences")
        print(self.user_email)
        if not success:
            print("Warning: Failed to save email to preferences file.")
```
d. Again in `do_filepath`, from:
```
    def do_filepath(self, arg):
        '''User specifies the project's filepath'''
        self.update_history(self.cmd_history, "filepath")

        
        while True:
            prompt (… etc)
```
To:
```
    def do_filepath(self, arg):
        '''User specifies the project's filepath'''
        self.update_history(self.cmd_history, "filepath")

        # Show current filepath if exists
        current_path = self.preferences.get_project_filepath()
        if current_path:
            print(f"Current filepath: {current_path}")
        
        while True:
            prompt (… etc)
```
